### ISSUE TITLE
wireshark-dissector: crash wmem_alloc: assertion failed

### ISSUE THREAD

**Issue by ChillerDragon** [heinrich5991/libtw2#73]
[test.zip](https://github.com/heinrich5991/libtw2/files/10950426/test.zip)

^ pcap that causes the crash

crash log:
```
$ wireshark test.pcap
ERROR:/usr/src/debug/wireshark/wireshark-4.0.4/wsutil/wmem/wmem_core.c:38:wmem_alloc: assertion failed: (allocator->in_scope)
Bail out! ERROR:/usr/src/debug/wireshark/wireshark-4.0.4/wsutil/wmem/wmem_core.c:38:wmem_alloc: assertion failed: (allocator->in_scope)
Aborted (core dumped)
```

My libtw2 version is self compiled at commit 5b51501942b8333d43bb9a555a48f1fbad59fb4a

```
$ wireshark --version
Wireshark 4.0.4 (Git v4.0.4 packaged as 4.0.4-1).
```

**Comment by heinrich5991** [comment]
Can reproduce the crash.

```
#0  __pthread_kill_implementation (threadid=<optimized out>, signo=signo@entry=6, no_tid=no_tid@entry=0) at pthread_kill.c:44
#1  0x00007fffee66d953 in __pthread_kill_internal (signo=6, threadid=<optimized out>) at pthread_kill.c:78
#2  0x00007fffee61eea8 in __GI_raise (sig=sig@entry=6) at ../sysdeps/posix/raise.c:26
#3  0x00007fffee60853d in __GI_abort () at abort.c:79
#4  0x00007fffef77406a in ?? () from /usr/lib/libglib-2.0.so.0
#5  0x00007fffef7cfa90 in g_assertion_message_expr () from /usr/lib/libglib-2.0.so.0
#6  0x00007fffef8a391e in wmem_alloc (allocator=<optimized out>, size=<optimized out>) at /usr/src/debug/wireshark/wireshark-4.0.4/wsutil/wmem/wmem_core.c:38
#7  0x00007fff9d7497f0 in wireshark_dissector::tw::dissect_impl (tvb=0x5555581f4de0, pinfo=0x7fffffffde38, ttree=0x0) at wireshark-dissector/src/tw.rs:264
#8  0x00007fff9d746a84 in wireshark_dissector::tw::dissect_heur (tvb=0x5555581f4de0, pinfo=0x7fffffffde38, ttree=0x0, _data=0x0) at wireshark-dissector/src/tw.rs:72
#9  0x00007ffff3803c70 in dissector_try_heuristic (sub_dissectors=0x7fff8c335f90, tvb=0x5555581f4de0, pinfo=0x7fffffffde38, tree=0x0, heur_dtbl_entry=0x7fffffffca90, data=0x0)
    at /usr/src/debug/wireshark/wireshark-4.0.4/epan/packet.c:2947
#10 0x00007ffff314764a in decode_udp_ports (tvb=<optimized out>, offset=<optimized out>, pinfo=0x7fffffffde38, udp_tree=<optimized out>, uh_sport=<optimized out>, 
    uh_dport=<optimized out>, uh_ulen=44) at /usr/src/debug/wireshark/wireshark-4.0.4/epan/dissectors/packet-udp.c:692
#11 0x00007ffff31486c2 in dissect (tvb=tvb@entry=0x5555581f4d90, pinfo=0x7fffffffde38, tree=<optimized out>, ip_proto=ip_proto@entry=17)
    at /usr/src/debug/wireshark/wireshark-4.0.4/epan/dissectors/packet-udp.c:1251
#12 0x00007ffff31490f2 in dissect_udp (tvb=0x5555581f4d90, pinfo=<optimized out>, tree=<optimized out>, data=<optimized out>)
    at /usr/src/debug/wireshark/wireshark-4.0.4/epan/dissectors/packet-udp.c:1257
#13 0x00007ffff3800753 in call_dissector_through_handle (handle=handle@entry=0x55555773af70, tvb=tvb@entry=0x5555581f4d90, pinfo=pinfo@entry=0x7fffffffde38, tree=tree@entry=0x0, 
    data=data@entry=0x555559911780) at /usr/src/debug/wireshark/wireshark-4.0.4/epan/packet.c:828
#14 0x00007ffff3801147 in call_dissector_work (handle=0x55555773af70, tvb=tvb@entry=0x5555581f4d90, pinfo=pinfo@entry=0x7fffffffde38, tree=tree@entry=0x0, 
    add_proto_name=add_proto_name@entry=1, data=data@entry=0x555559911780) at /usr/src/debug/wireshark/wireshark-4.0.4/epan/packet.c:919
#15 0x00007ffff38012b3 in dissector_try_uint_new (sub_dissectors=<optimized out>, uint_val=17, tvb=0x5555581f4d90, pinfo=0x7fffffffde38, tree=0x0, add_proto_name=1, data=0x555559911780)
    at /usr/src/debug/wireshark/wireshark-4.0.4/epan/packet.c:1522
#16 0x00007ffff2c45109 in ip_try_dissect (heur_first=<optimized out>, nxt=17, tvb=tvb@entry=0x5555581f4d90, pinfo=pinfo@entry=0x7fffffffde38, tree=tree@entry=0x0, 
    iph=iph@entry=0x555559911780) at /usr/src/debug/wireshark/wireshark-4.0.4/epan/dissectors/packet-ip.c:1822
#17 0x00007ffff2c4640f in dissect_ip_v4 (tvb=0x5555581f4d40, pinfo=<optimized out>, parent_tree=0x0, data=<optimized out>)
    at /usr/src/debug/wireshark/wireshark-4.0.4/epan/dissectors/packet-ip.c:2328
#18 0x00007ffff3800753 in call_dissector_through_handle (handle=handle@entry=0x55555789c1e0, tvb=tvb@entry=0x5555581f4d40, pinfo=pinfo@entry=0x7fffffffde38, tree=tree@entry=0x0, 
    data=data@entry=0x0) at /usr/src/debug/wireshark/wireshark-4.0.4/epan/packet.c:828
#19 0x00007ffff3801147 in call_dissector_work (handle=0x55555789c1e0, tvb=tvb@entry=0x5555581f4d40, pinfo=pinfo@entry=0x7fffffffde38, tree=tree@entry=0x0, 
    add_proto_name=add_proto_name@entry=1, data=data@entry=0x0) at /usr/src/debug/wireshark/wireshark-4.0.4/epan/packet.c:919
#20 0x00007ffff38012b3 in dissector_try_uint_new (sub_dissectors=<optimized out>, uint_val=2048, tvb=0x5555581f4d40, pinfo=0x7fffffffde38, tree=0x0, add_proto_name=add_proto_name@entry=1, 
    data=0x0) at /usr/src/debug/wireshark/wireshark-4.0.4/epan/packet.c:1522
#21 0x00007ffff3801326 in dissector_try_uint (sub_dissectors=<optimized out>, uint_val=<optimized out>, tvb=<optimized out>, pinfo=<optimized out>, tree=<optimized out>)
    at /usr/src/debug/wireshark/wireshark-4.0.4/epan/packet.c:1546
#22 0x00007ffff2a9fa7c in dissect_ethertype (tvb=0x5555598d6a40, pinfo=0x7fffffffde38, tree=0x0, data=0x7fffffffd190)
    at /usr/src/debug/wireshark/wireshark-4.0.4/epan/dissectors/packet-ethertype.c:296
#23 0x00007ffff3800753 in call_dissector_through_handle (handle=handle@entry=0x5555576ca6c0, tvb=tvb@entry=0x5555598d6a40, pinfo=pinfo@entry=0x7fffffffde38, tree=tree@entry=0x0, 
    data=data@entry=0x7fffffffd190) at /usr/src/debug/wireshark/wireshark-4.0.4/epan/packet.c:828
#24 0x00007ffff3801147 in call_dissector_work (handle=0x5555576ca6c0, tvb=tvb@entry=0x5555598d6a40, pinfo=pinfo@entry=0x7fffffffde38, tree=tree@entry=0x0, 
    add_proto_name=add_proto_name@entry=1, data=0x7fffffffd190) at /usr/src/debug/wireshark/wireshark-4.0.4/epan/packet.c:919
#25 0x00007ffff38014d7 in call_dissector_only (handle=<optimized out>, tvb=tvb@entry=0x5555598d6a40, pinfo=pinfo@entry=0x7fffffffde38, tree=tree@entry=0x0, data=<optimized out>)
    at /usr/src/debug/wireshark/wireshark-4.0.4/epan/packet.c:3398
#26 0x00007ffff38017a7 in call_dissector_with_data (handle=<optimized out>, tvb=0x5555598d6a40, pinfo=0x7fffffffde38, tree=0x0, data=<optimized out>)
    at /usr/src/debug/wireshark/wireshark-4.0.4/epan/packet.c:3411
#27 0x00007ffff2a9c098 in dissect_eth_common (tvb=tvb@entry=0x5555598d6a40, pinfo=pinfo@entry=0x7fffffffde38, parent_tree=parent_tree@entry=0x0, fcs_len=fcs_len@entry=-1)
    at /usr/src/debug/wireshark/wireshark-4.0.4/epan/dissectors/packet-eth.c:595
#28 0x00007ffff2a9ca8a in dissect_eth (tvb=0x5555598d6a40, pinfo=0x7fffffffde38, tree=0x0, data=<optimized out>)
    at /usr/src/debug/wireshark/wireshark-4.0.4/epan/dissectors/packet-eth.c:901
#29 0x00007ffff3800753 in call_dissector_through_handle (handle=handle@entry=0x555557869720, tvb=tvb@entry=0x5555598d6a40, pinfo=pinfo@entry=0x7fffffffde38, tree=tree@entry=0x0, 
    data=data@entry=0x7fffffffe188) at /usr/src/debug/wireshark/wireshark-4.0.4/epan/packet.c:828
#30 0x00007ffff3801147 in call_dissector_work (handle=0x555557869720, tvb=0x5555598d6a40, pinfo=0x7fffffffde38, tree=0x0, add_proto_name=1, data=0x7fffffffe188)
    at /usr/src/debug/wireshark/wireshark-4.0.4/epan/packet.c:919
--Type <RET> for more, q to quit, c to continue without paging--c
#31 0x00007ffff2ad07fd in dissect_frame (tvb=0x5555598d6a40, pinfo=<optimized out>, parent_tree=0x0, data=0x7fffffffd8a0)
    at /usr/src/debug/wireshark/wireshark-4.0.4/epan/dissectors/packet-frame.c:947
#32 0x00007ffff3800753 in call_dissector_through_handle (handle=handle@entry=0x5555576cdc60, tvb=tvb@entry=0x5555598d6a40, pinfo=pinfo@entry=0x7fffffffde38, tree=tree@entry=0x0, 
    data=data@entry=0x7fffffffd8a0) at /usr/src/debug/wireshark/wireshark-4.0.4/epan/packet.c:828
#33 0x00007ffff3801147 in call_dissector_work (handle=0x5555576cdc60, tvb=tvb@entry=0x5555598d6a40, pinfo=pinfo@entry=0x7fffffffde38, tree=tree@entry=0x0, 
    add_proto_name=add_proto_name@entry=1, data=0x7fffffffd8a0) at /usr/src/debug/wireshark/wireshark-4.0.4/epan/packet.c:919
#34 0x00007ffff38014d7 in call_dissector_only (handle=<optimized out>, tvb=tvb@entry=0x5555598d6a40, pinfo=pinfo@entry=0x7fffffffde38, tree=tree@entry=0x0, data=<optimized out>)
    at /usr/src/debug/wireshark/wireshark-4.0.4/epan/packet.c:3398
#35 0x00007ffff38017a7 in call_dissector_with_data (handle=<optimized out>, tvb=0x5555598d6a40, pinfo=0x7fffffffde38, tree=0x0, data=<optimized out>)
    at /usr/src/debug/wireshark/wireshark-4.0.4/epan/packet.c:3411
#36 0x00007ffff3801cd0 in dissect_record (edt=edt@entry=0x7fffffffde20, file_type_subtype=file_type_subtype@entry=1, rec=rec@entry=0x7fffffffe150, tvb=tvb@entry=0x5555598d6a40, 
    fd=fd@entry=0x5555598f91e0, cinfo=cinfo@entry=0x0) at /usr/src/debug/wireshark/wireshark-4.0.4/epan/packet.c:632
#37 0x00007ffff3801e77 in epan_dissect_run_with_taps (edt=0x7fffffffde20, file_type_subtype=1, rec=0x7fffffffe150, tvb=0x5555598d6a40, fd=0x5555598f91e0, cinfo=0x0)
    at /usr/src/debug/wireshark/wireshark-4.0.4/epan/epan.c:638
#38 0x000055555597c34c in add_packet_to_packet_list (fdata=0x5555598f91e0, cf=0x555555d98b20 <cfile>, edt=0x7fffffffde20, dfcode=0x0, cinfo=0x0, rec=0x7fffffffe150, buf=0x7fffffffde00, 
    add_to_packet_list=1) at /usr/src/debug/wireshark/wireshark-4.0.4/file.c:1192
#39 0x000055555597c7fe in read_record (cf=cf@entry=0x555555d98b20 <cfile>, rec=rec@entry=0x7fffffffe150, buf=buf@entry=0x7fffffffde00, dfcode=0x0, edt=edt@entry=0x7fffffffde20, 
    cinfo=cinfo@entry=0x0, offset=24) at /usr/src/debug/wireshark/wireshark-4.0.4/file.c:1288
#40 0x0000555555984b47 in cf_read (cf=0x555555d98b20 <cfile>, reloading=0) at /usr/src/debug/wireshark/wireshark-4.0.4/file.c:640
#41 0x0000555555879af1 in WiresharkMainWindow::openCaptureFile (this=0x555555ea6fa0, cf_path=..., read_filter=..., type=<optimized out>, is_tempfile=0)
    at /usr/src/debug/wireshark/wireshark-4.0.4/ui/qt/capture_file.cpp:239
#42 0x000055555565f643 in main (argc=<optimized out>, qt_argv=<optimized out>) at /usr/src/debug/wireshark/wireshark-4.0.4/ui/qt/main.cpp:941
```

**Comment by heinrich5991** [comment]
It's apparently a Wireshark bug. Don't use Wireshark 4.0.4 to fix the problem, either downgrade to 4.0.3 or upgrade to 4.0.5 once it's available.

**Comment by heinrich5991** [comment]
If you're on Arch Linux, you can downgrade with this command (assuming you haven't cleared the caches):
```
sudo pacman -U /var/cache/pacman/pkg/wireshark-{cli,qt}-4.0.3-1-x86_64.pkg.tar.zst
```
